internal/repository/master_index.go

```diff
@@ -237,6 +237,27 @@ func (mi *MasterIndex) Each(ctx context.Context) <-chan restic.PackedBlob {
 	return ch
 }
 
+func (mi *MasterIndex) MergeFinalIndexes() {
+	mi.idxMutex.Lock()
+	defer mi.idxMutex.Unlock()
+
+	var firstFinalIndex *Index
```

Comment:
I think we have a rather subtle/obscure race condition between multiple `saveIndex` calls here. Two `saveIndex` calls could be triggered in parallel (just upload two packs in parallel with over 50k small blobs). Then the first one to finish will call `MergeFinalIndexes` which processes _all_ final indexes. Now, if the second call to `saveIndex` completes first, then the `firstFinalIndex` index will actually belong to the first `saveIndex` run. And it's probably not a good idea to modify the index while it waits to be uploaded (For this to actually be a problem the second saveIndex run would have to complete before the first one serialized the first index. And that's probably quite unlikely, but not impossible). I hope that my explanation makes sense.

A really simple way to solve this would be to prime the MasterIndex with an empty but finalized initial Index. That would ensure that the first index is indeed final and also prevents the race condition as indexes would all be merged into that placeholder index.

Reply:
Ah, you are right. I was thinking about adding a "main" index into MasterIndex.
I also think that your scenario can only occur during the first initial backup, i.e. when starting with a completely empty repository. But we should be prepared for this case...

Reply:
I changed NewMasterIndex() to always generate an empty final index as first list element.